File: android/src/main/java/com/paulingalls/realtimeaudio/RealtimeAudioPlayerModule.kt
package com.paulingalls.realtimeaudio

import AudioFormatSettings
import RealtimeAudioPlayer
import RealtimeAudioPlayerDelegate
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import getAndroidColor
import mapAudioEncodingToFormat
import mapChannelCountToOutputFormat

class RealtimeAudioPlayerModule : Module() {
    var hasListeners = false

    override fun definition() = ModuleDefinition {
        Name("RealtimeAudioPlayer")
        Events("onPlaybackStarted", "onPlaybackStopped")

        OnStartObserving {
            hasListeners = true
        }

        OnStopObserving {
            hasListeners = false
        }

        Class("RealtimeAudioPlayer", RealtimeAudioPlayer::class) {
            Constructor { format: AudioFormatSettings ->
                return@Constructor RealtimeAudioPlayer(
                    format.sampleRate,
                    mapChannelCountToOutputFormat(format.channelCount),
                    mapAudioEncodingToFormat(format.encoding)
                ).apply {
                    delegate = RealtimeEventDelegate(this@RealtimeAudioPlayerModule)
                }
            }

            AsyncFunction("addBuffer") { player: RealtimeAudioPlayer, base64String: String ->
                player.addBuffer(base64String)
            }
            AsyncFunction("pause") { player: RealtimeAudioPlayer ->
                player.pausePlayback()
            }
            AsyncFunction("resume") { player: RealtimeAudioPlayer ->
                player.resumePlayback()
            }
            AsyncFunction("stop") { player: RealtimeAudioPlayer ->
                player.stopPlayback()
            }
        }

        View(RealtimeAudioPlayerView::class) {
            Events("onPlaybackStarted", "onPlaybackStopped")

            Prop("waveformColor") { view: RealtimeAudioPlayerView, hexColor: String ->
                view.setVisualizationColor(getAndroidColor(hexColor))
            }

            Prop("audioFormat") { view: RealtimeAudioPlayerView, format: AudioFormatSettings ->
                view.setAudioFormat(
                    format.sampleRate,
                    mapChannelCountToOutputFormat(format.channelCount),
                    mapAudioEncodingToFormat(format.encoding)
                )
            }

            AsyncFunction("addBuffer") { view: RealtimeAudioPlayerView, base64String: String ->
                view.addAudioBuffer(base64String)
            }
            AsyncFunction("pause") { view: RealtimeAudioPlayerView ->
                view.pausePlayback()
            }
            AsyncFunction("resume") { view: RealtimeAudioPlayerView ->
                view.resumePlayback()
            }
            AsyncFunction("stop") { view: RealtimeAudioPlayerView ->
                view.stopPlayback()
            }
        }
    }

    class RealtimeEventDelegate(
        private val module: RealtimeAudioPlayerModule
    ) : RealtimeAudioPlayerDelegate {

        override fun playbackStarted() {
            if (module.hasListeners) {
                module.sendEvent("onPlaybackStarted")
            }
        }

        override fun playbackStopped() {
            if (module.hasListeners) {
                module.sendEvent("onPlaybackStopped")
            }
        }

        override fun bufferReady(buffer: ByteArray) {
        }
    }

}

File: ios/RealtimeAudioPlayerModule.swift
import AVFoundation
import ExpoModulesCore

public class RealtimeAudioPlayerModule:
  Module, RealtimeAudioPlayerDelegate {
  var hasListeners: Bool = false
  
  public func definition() -> ModuleDefinition {
    Name("RealtimeAudioPlayer")
    
    Events("onPlaybackStarted", "onPlaybackStopped")
    
    OnStartObserving {
      hasListeners = true
    }
    
    OnStopObserving {
      hasListeners = false
    }
    
    Class(RealtimeAudioPlayer.self) {
      Constructor { (audioFormat: AudioFormatSettings) -> RealtimeAudioPlayer in
        let player: RealtimeAudioPlayer = RealtimeAudioPlayer(sampleRate: audioFormat.sampleRate,
                                                              commonFormat: getCommonFormat(audioFormat.encoding),
                                                              channels: audioFormat.channelCount)!
        player.delegate = self
        return player
      }
      
      // Functions
      AsyncFunction("addBuffer") { (
        player: RealtimeAudioPlayer,
        base64String: String
      ) in
        player.addBuffer(base64String)
      }
      
      AsyncFunction("resume") { (player: RealtimeAudioPlayer) in
        player.resume()
      }
      
      AsyncFunction("pause") { (player: RealtimeAudioPlayer) in
        player.pause()
      }
      
      AsyncFunction("stop") { (player: RealtimeAudioPlayer) in
        player.stop()
      }
    }
    
    View(RealtimeAudioPlayerView.self) {
      Events("onPlaybackStarted", "onPlaybackStopped")
      
      // Props
      Prop("waveformColor") { (
        view: RealtimeAudioPlayerView,
        hexColor: UIColor
      ) in
        view.setWaveformColor(hexColor)
      }
      
      Prop("audioFormat") { (
        view: RealtimeAudioPlayerView,
        format: AudioFormatSettings
      ) in
        view.setAudioFormat(
          sampleRate: format.sampleRate,
          commonFormat: self.getCommonFormat(format.encoding),
          channels: format.channelCount
        )
      }
      
      // Functions
      AsyncFunction("addBuffer") { (
        view: RealtimeAudioPlayerView,
        base64String: String
      ) in
        view.addBuffer(base64String)
      }
      
      AsyncFunction("resume") { (view: RealtimeAudioPlayerView) in
        view.resume()
      }
      
      AsyncFunction("pause") { (view: RealtimeAudioPlayerView) in
        view.pause()
      }
      
      AsyncFunction("stop") { (view: RealtimeAudioPlayerView) in
        view.stop()
      }
    }
  }
  
  func audioPlayerDidStartPlaying() {
    if hasListeners { sendEvent("onPlaybackStarted") }
  }
  
  func audioPlayerDidStopPlaying() {
    if hasListeners { sendEvent("onPlaybackStopped") }
  }
  
  func audioPlayerBufferDidBecomeAvailable(_ buffer: AVAudioPCMBuffer) {
  }
  
  private func getCommonFormat(_ encoding: AudioEncoding) -> AVAudioCommonFormat {
    switch encoding {
    case .pcm16bitInteger:
      return .pcmFormatInt16
    case .pcm32bitInteger:
      return .pcmFormatInt32
    case .pcm32bitFloat:
      return .pcmFormatFloat32
    case .pcm64bitFloat:
      return .pcmFormatFloat64
    }
  }
}
